### DISCUSSION TITLE
macos crash on catalina with missing swiftui symbol

### DISCUSSION THREAD

**Discussion by jaredh159** [pointfreeco/swift-composable-architecture#2351] [Q&A]
Hi there, thanks so much for TCA, really appreciate all your work.

I've got a macos app that was already in production with a prerelease of 1.0, and now has been converted to `1.0.0`. in a recent release, I've been trying to add support for an older OS, namely Catalina 10.15 (I already support everything from Big Sur and later).

In testing the new Catalina compat, I'm running into a strange crash on launch that seems to be related to TCA, and I'm struggling to make heads or tails of it. The crash is due to a missing SwiftUI symbol.  Here's the only interesting part of the crash report:

```
Symbol not found: _$s7SwiftUI4ViewPAAE15fullScreenCover4item9onDismiss7contentQrAA7BindingVyqd__SgG_yycSgqd_0_qd__cts12IdentifiableRd__AaBRd_0_r0_lFQOMQ
```
running `swift demangle` on that symbol produces:

```
opaque type descriptor for <<opaque return type of (extension in SwiftUI):SwiftUI.View.fullScreenCover<A, B where A1: Swift.Identifiable, B1: SwiftUI.View>(item: SwiftUI.Binding<A1?>, onDismiss: (() -> ())?, content: (A1) -> B1) -> some>>
```

Why do I think it's related to TCA? Because my app uses [zero SwiftUI](https://github.com/pointfreeco/swift-composable-architecture/discussions/2031), so the only swiftui code that should be referenced (I think) should be coming from TCA.

The strange thing is, [searching pointfree repos for `fullScreenCover`](https://github.com/search?q=org%3Apointfreeco%20fullScreenCover&type=code) turns up some stuff, but the [relevant lines](https://github.com/pointfreeco/swift-composable-architecture/blob/c895f87166c5b8b381af35cb1bde15870aee08bf/Sources/ComposableArchitecture/SwiftUI/FullScreenCover.swift#L20) seem to be correctly annotated with `@available(macOS, unavailable)`.

It's also add, that I'm not getting the crash on anything macos 11.0 or later, all of which don't support `fullScreenCover` either (I think).

Admittedly, I'm not a veteran macos or ios developer, so maybe I'm missing something stupid here?  Can anyone help me understand what's happening here? 

Does anyone have a TCA app building and running successfully on Catalina? Is that something you have a sample app for that could be tested for the crash?

Thanks for any help!

<details>
<summary>expand for more of the crash report...</summary>

```
Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Termination Reason:    DYLD, [0x4] Symbol missing

Application Specific Information:
dyld: launch, loading dependent libraries

Dyld Error Message:
  Symbol not found: _$s7SwiftUI4ViewPAAE15fullScreenCover4item9onDismiss7contentQrAA7BindingVyqd__SgG_yycSgqd_0_qd__cts12IdentifiableRd__AaBRd_0_r0_lFQOMQ
  Referenced from: /Applications/MyApp.app/Contents/MacOS/MyApp
  Expected in: /System/Library/Frameworks/SwiftUI.framework/Versions/A/SwiftUI
 in /Applications/MyApp.app/Contents/MacOS/MyApp
```

</details>

**Comment by mbrandonw** [comment]
The demangled symbol you shared does not match the symbol you [linked to in TCA](https://github.com/pointfreeco/swift-composable-architecture/blob/c895f87166c5b8b381af35cb1bde15870aee08bf/Sources/ComposableArchitecture/SwiftUI/FullScreenCover.swift#L20). The TCA `fullscreenCover` has 3 generics and deals with `Store`s, whereas your demangled symbol is much simpler. In fact, I'm pretty sure it is just the plain, [vanilla SwiftUI `fullscreenCover`](https://developer.apple.com/documentation/swiftui/view/fullscreencover(item:ondismiss:content:)).

What was the previous version of TCA your project was building without crashing?

**Reply by jaredh159** [reply]
Thanks for the prompt reply.

To clarify, it _builds_ just fine, rather the notarized app crashes on application _launch_.

The previous version of my TCA project did not target 10.15, so this is unexplored territory for me. I'm trying to add support for 10.15 and seeing this crash now.

After submitting this, I just found:

https://github.com/pointfreeco/swiftui-navigation/discussions/100

Which seems to be the exact same thing. You said it worked for you in that thread, but did you actually **launch a built app on a computer running Catalina?** He seemed to have a test project showing the crash, but there were no more replies to that thread after he posted the test project.

You're probably right about the symbol not matching the TCA code.  But if my app uses zero SwiftUI, where/why would a vanilla swiftui method be getting imported/called?

**Reply by mbrandonw** [reply]
Sorry, I misspoke. I understand it's a runtime crash. 

> but did you actually **launch a built app on a computer running Catalina?**

I don't have access to a Catalina machine to test this. Did you run his project and experience the same crash?

**Reply by jaredh159** [reply]
I haven't run his project yet, but I'm going to try. I don't have my own Catalina computer, I've been borrowing my nieces off and on and don't have it at the moment. But I'm trying to setup a dual-boot to Catalina right now, or borrow her computer again today.  ...sigh, the world of macos development.  i'll try to report back what i find repro-ing his test case.

**Reply by jaredh159** [reply]
ok, i have my hands on a laptop with Catalina now, and have some more info:

1. the test project from the other thread _does_ immediately crash with the same error on Catalina
2. i realized i had a few `import SwiftUI` statements in my codebase, but they were all just to get symbols that I could also get from `AppKit` (like `NSWindow`), so i removed those, built, notarized and tested.  ...but sadly, the crash persists.

**Reply by mbrandonw** [reply]
It seems like the crash should happen with just SwiftUINavigation. No need for TCA at all. I'm wondering if you can slowly remove things from SwiftUINavigation to get a minimal reproduction? Is it just the presence of the `fullScreenCover` method that is marked as `@available(macOS, unavailable)` and calls `fullScreenCover` under the hood?

**Reply by jaredh159** [reply]
@mbrandonw ok, one final piece of interesting info.

i thought some more about how you said the symbol didn't match the TCA code. you're definitely right, but the TCA code calls out to the vanilla swiftui method that is missing it's symbol [here](https://github.com/pointfreeco/swift-composable-architecture/blob/c895f87166c5b8b381af35cb1bde15870aee08bf/Sources/ComposableArchitecture/SwiftUI/FullScreenCover.swift#L26). 

like i said, i'm not a macos/ios vet, so instead of bothering you wondering about if that was still somehow the cause, i did an experiment. i forked TCA, [commented out that entire file](https://github.com/jaredh159/swift-composable-architecture/commit/639f2d1994c997c44680c6599c4f9a58b88f7960), depended on my fork, rebuilt, re-notarized, and tested.

and the crash is **gone**!

i have no idea why that would be. like i said, you seem to be correctly annotating it as not available for macos. but there it is. ¯\\_(ツ)_/¯

it honestly feels maybe like more of a xcode bug than something you guys are doing wrong. but, i have a hunch that _nothing_ built with TCA 1.0.0 will launch without a crash on Catalina right now. 

i have this Catalina computer for a couple days. is it possibly something easy for you to do to build a sample TCA app for macos with the deployment target set to 10.15, and have it notarized?  if you could do that and send me a `.zip` or `.dmg` i'd be happy to test it.

**Reply by jaredh159** [reply]
oh, didn't see your response until i finished my last response. i think you were thinking along the same lines, although i was able to fix it by commenting out just one file in TCA.

**Comment by jaredh159** [comment]
@mbrandonw ok, so i purchased an old laptop to do Catalina testing (for $70 🤣 ), and I dug into this issue again.  here's a couple more bits of into.

As I suspected, it seems that _any macOS app importing TCA is crashing on Catalina now_. To test this, I made the simplest TCA app I could (the humble counter app, of course!), and verified that I get the same crash on Catalina. The [github repo is here](https://github.com/jaredh159/catalina-compat-tca-counter).

As a contrast, here's just a hello-world SwiftUI app without TCA, that [doesn't crash on catalina](https://github.com/jaredh159/catalina-compat-swiftui-hello-world).

But, if I add this just this code block to the **non**-TCA swiftui app:

```swift
public extension View {
  @available(iOS 14, tvOS 14, watchOS 7, *)
  @available(macOS, unavailable)
  func testAvailableCrash() -> some View {
    self.fullScreenCover(isPresented: .constant(true)) {
      Text("Unreachable!")
    }
  }
}
```

...then we're in crash city again. This is basically the minimal code I could figure out to reproduce the fact that for some reason it seems like the `@available(macOS, unavailable)` annotation is not preventing that codepath from running on Catalina, resulting in the missing symbol crash.  That code is basically simulating the [one file in TCA that causes the crash](https://github.com/pointfreeco/swift-composable-architecture/blob/c895f87166c5b8b381af35cb1bde15870aee08bf/Sources/ComposableArchitecture/SwiftUI/FullScreenCover.swift#L26).

I'd love to get to the true bottom of this. Do you have any more knowledge about how those `@available` checks are supposed to work, or if there is some sort of alternate compiler/availability directive that would allow this to still work for the platforms that support it, without breaking Catalina compat?

I'd really hate for you guys to drop Catalina support, I hope you're not considering that! 🤞 (on that note, now that I've got a dedicated Catalina test computer, if we can get this resolved some way in the core library, I'd be happy to do some smoke-testing for you guys whenever you are about to tag a new release, to ensure Catalina is hunky dory -- let me know if that interests you).

**Reply by mbrandonw** [reply]
Hi @jaredh159, thanks for investigating more!

Since you have such a small reproducible test case it sounds like it's time to take the bug to Apple. It almost certainly seems like an issue with Swift. I would recommend opening an issue on the Swift repo _and_ starting a Swift forum discussion for visibility. If there is something simple we can do in the library to work around we would be happy to.

I guess one idea that comes to mind… would it suddenly work if you wrapped the helper that uses `@available(macOS, unavailable)` in `#if !os(macOS) … #endif`? That way it just doesn't compile at all?

**Reply by jaredh159** [reply]
thanks for the reply. i tried a couple `if #available` directives, but couldn't get one that worked. i didn't try what you proposed there, but unless I'm mistaken, that wouldn't be an option, right? Wouldn't that prevent any code from being ever emitted into the binary within the #if block, effectively making the code also not available for iOS in cross-platform apps? I was thinking we needed some kind of runtime codepath protection like `@available` or `if #available`, not something that started with `#if` -- though I'm not 100% clear that I perfectly understand the differences between those 3 flavors.

i thought about a swift forum post, but i thought they generally discourage posts about apple platforms, is that not correct?  i'm open to opening an issue over there and on the swift repo as well. i did post over on the developer forums -- sometimes that's a black hole, but i've sometimes gotten good input from Eskimo at least:  https://developer.apple.com/forums/thread/735910

**Reply by mbrandonw** [reply]
> i didn't try what you proposed there, but unless I'm mistaken, that wouldn't be an option, right? Wouldn't that prevent any code from being ever emitted into the binary within the #if block, effectively making the code also not available for iOS in cross-platform apps? I was thinking we needed some kind of runtime codepath protection like `@available` or `if #available`, not something that started with `#if` -- though I'm not 100% clear that I perfectly understand the differences between those 3 flavors.

As far as I can tell, `@available(macOS, unavailable)` is roughly equivalent to `#if !os(macOS)`, outside of ergonomics. For complex `@available` chaining it of course easier to use the annotation than a bunch of `#if`s, but that isn't a concern here. If there is a Swift/macOS bug that doesn't like to compile that code on Mac, then let's just omit the code from Mac entirely.

Can you check if that gets your app compiling? If so we can make that change to the library very easily.

>  thought about a swift forum post, but i thought they generally discourage posts about apple platforms, is that not correct?

This does feel like an actual bug in the Swift compiler, though it may be entangled with the macOS runtime. Really hard to say!

I do feel that this kind of question is appropriate for the Swift forums if you don't get any traction in the Apple developer forums. It feels that the questions they most want out of the Swift forums are things having to specifically do with Apple framework APIs. And this can be reproduced in very vanilla Swift code. You don't even need the `extension View` I imagine. Seems like it could just be any plain function with the `@available` attribute.

**Reply by jaredh159** [reply]
I tested with the code changed to this:

```swift
#if !os(macOS)
  public extension View {
    @available(iOS 14, tvOS 14, watchOS 7, *)
    func testAvailableCrash() -> some View {
      self.fullScreenCover(isPresented: .constant(true)) {
        Text("Unreachable!")
      }
    }
  }
#endif
```

And it does prevent the crash on Catalina, but I'm still not sure that's correct. Based on [this post](https://forums.swift.org/t/if-vs-available-vs-if-available/40266/2) it seems like that would be testing the **build machine** which would always be macOS, since it's evaluated at _compile_ time. at compile time, you'd always be on macOS (unless you're on Linux building a server app or something like that). or am I just majorly misunderstanding that post?

I opened an issue on the Swift repo: https://github.com/apple/swift/issues/68010 -- I guess i'll go ahead and try one on the Swift forums for good measure as well.

**Reply by mbrandonw** [reply]
> And it does prevent the crash on Catalina, but I'm still not sure that's correct. Based on [this post](https://forums.swift.org/t/if-vs-available-vs-if-available/40266/2) it seems like that would be testing the **build machine** which would always be macOS, since it's evaluated at _compile_ time. at compile time, you'd always be on macOS (unless you're on Linux building a server app or something like that). or am I just majorly misunderstanding that post?

It's not the platform you are building on. Otherwise things like `#if os(watchOS)` wouldn't make much sense since you are never building on a watch. Instead it refers to the platform you are building _for_, i.e. the platform you have selected at the top of the Xcode window. We already have a few `#if os(…)` scattered throughout the library for just this kind of thing.

**Reply by jshier** [reply]
> As far as I can tell, `@available(macOS, unavailable)` is roughly equivalent to `#if !os(macOS)`, outside of ergonomics.

That's not really true. `#if` is a compiler directive, so it's a compile-time only construct. `#if !os(macOS)` means the API won't exist at all when built for a macOS deployment target. `@available` (in combination with `if #available`) is a *runtime* construct. The code is still built and part of the executable but the compiler checks your usage at build time. If this runtime check doesn't work for some reason, you may still see it executed despite `@available(macOS, unavailable)`. You can see this on godbolt.org (using Linux instead of macOS).

```swift
@available(Linux, unavailable)
func square(n: Int) -> Int {
    return n * n
}
```

With this version, the code is still generated, you just can't call it from other function (thought godbolt is letting me, odd). If you instead use `#if !os(Linux)` you'll see the function isn't built at all. It just doesn't exist.

I don't know the root cause of this bug. If it's a compiler issue, you may be able to get it fixed. But since it only occurs on Catalina it seems likely it's a runtime bug of some kind (or perhaps a compiler bug that only shows up on older runtimes), and that won't be fixed since they aren't updating Catalina anymore, unless the solution is only in the compiler.

**Reply by mbrandonw** [reply]
I agree with your detailed explanation, but I said _roughly equivalent_ specifically to not go into those details. At the end of the day you cannot call the function and it doesn't even autocomplete in Xcode. So we might as well omit it entirely from the library if it is causing a crash on some platform.

**Reply by jaredh159** [reply]
ok, thanks, that makes sense, i was just slowly coming to that conclusion from some more googling. the objc version of the same check is `#ifdef TARGET_OS_TV` which includes the helpful word "target".  and i forgot that for a multi-platform app, you're going to be building more than once and emitting different app bundles.

interestingly, i found this (sadly unanswered) post while i was starting to work on a swift form post: https://forums.swift.org/t/available-unavailable-api-unexpected-behavior/38784

whipped up a quick PR, if you'd be up for introducing the os check in this way. tested in my minimal repro, but i'll do one more test pulling in the fork and commit from that PR to double-check. here's the PR: https://github.com/pointfreeco/swift-composable-architecture/pull/2385

**Reply by jshier** [reply]
Sure, it just means every callsite will need to use the same construct, so if you wrap it in `#if !os(macOS)`, anyone with a cross platform build will need to do the same thing (unless they use some other construct to avoid it, like separate platform files).
